src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java

```diff
@@ -0,0 +1,439 @@
+
+            if (isPowered != this.isPowered) {
+                this.isPowered = isPowered;
+                writeCustomData(GregtechDataCodes.UPDATE_ENERGY, buf -> buf.writeBoolean(this.isPowered));
+            }
+        }
+    }
+
+    public boolean isPowered() {
+        return isPowered;
+    }
+
+    @Override
+    public void receiveCustomData(int dataId, @NotNull PacketBuffer buf) {
+        super.receiveCustomData(dataId, buf);
+        if (dataId == GregtechDataCodes.UPDATE_ENERGY) {
+            this.isPowered = buf.readBoolean();
+            scheduleRenderUpdate();
+        } else if (dataId == GregtechDataCodes.UPDATE_ENERGY_PER) {
+            this.energyConsumption = buf.readLong();
+        }
+    }
+
+    @Override
+    public void renderMetaTileEntity(CCRenderState renderState, Matrix4 translation, IVertexOperation[] pipeline) {
+        var front = isPowered() ?
+                Textures.QUANTUM_CONTROLLER_FRONT_ACTIVE :
+                Textures.QUANTUM_CONTROLLER_FRONT_INACTIVE;
+        var sides = isPowered() ?
+                Textures.QUANTUM_CONTROLLER_ACTIVE :
+                Textures.QUANTUM_CONTROLLER_INACTIVE;
+
+        var newPipeline = ArrayUtils.add(pipeline,
+                new ColourMultiplier(GTUtility.convertRGBtoOpaqueRGBA_CL(getPaintingColorForRendering())));
+
+        front.renderSided(getFrontFacing(), renderState, translation, newPipeline);
+
+        for (EnumFacing facing : EnumFacing.VALUES) {
+            if (facing == getFrontFacing()) continue;
+            sides.renderSided(facing, renderState, translation, newPipeline);
+        }
+    }
+
+    @Override
+    public Pair<TextureAtlasSprite, Integer> getParticleTexture() {
+        return Pair.of(Textures.QUANTUM_CONTROLLER_ACTIVE.getParticleSprite(), getPaintingColorForRendering());
+    }
+
+    @Override
+    protected boolean openGUIOnRightClick() {
+        return false; // todo use mui2 for ui?
+    }
+
+    @Override
+    public boolean isValidFrontFacing(EnumFacing facing) {
+        return true;
+    }
+
+    @SuppressWarnings("SameParameterValue")
+    private IQuantumStorage<?> getStorage(BlockPos pos, boolean rebuild) {
+        if (getWorld().isRemote) return null;
+        if (storageInstances.containsKey(pos)) {
+            WeakReference<IQuantumStorage<?>> storageRef = storageInstances.get(pos);
+            IQuantumStorage<?> storage = storageRef.get();
+            if (storage != null) {
+                return storage;
+            }
+        }
+        // need to make sure it still exists
+        MetaTileEntity mte = GTUtility.getMetaTileEntity(getWorld(), pos);
+        if (mte instanceof IQuantumStorage<?>storage) {
+            storageInstances.put(pos, new WeakReference<>(storage));
+            return storage;
+        } else if (rebuild) {
+            // need to remove and rebuild
+            storagePositions.remove(pos);
+            rebuildNetwork();
+            return null;
```

Comment:
Could rebuilding the network end up returning a valid storage in a new network? Do we have to take this into account?

Reply:
the `rebuild` paramater is always false, idk what seni intended for this